api/shared/condition_types.go

```diff
@@ -0,0 +1,155 @@
+	LastTransitionTime metav1.Time            `json:"lastTransitionTime,omitempty"`
+}
+
+// ConditionType - A summarizing name for a given condition
+type ConditionType string
+
+// ConditionReason - Why a particular condition is true, false or unknown
+type ConditionReason string
+
+// NewCondition - Create a new condition object
+func NewCondition(conditionType ConditionType, status corev1.ConditionStatus, reason ConditionReason, message string) Condition {
+	now := metav1.Time{Time: time.Now()}
+	condition := Condition{
+		Type:               conditionType,
+		Status:             status,
+		Reason:             reason,
+		Message:            message,
+		LastHeartbeatTime:  now,
+		LastTransitionTime: now,
+	}
+	return condition
+}
+
+// Set - Set a particular condition in a given condition list
+func (conditions *ConditionList) Set(conditionType ConditionType, status corev1.ConditionStatus, reason ConditionReason, message string) {
+	condition := conditions.Find(conditionType)
+
+	// If there isn't condition we want to change, add new one
+	if condition == nil {
+		condition := NewCondition(conditionType, status, reason, message)
+		*conditions = append(*conditions, condition)
+		return
+	}
+
+	now := metav1.Time{Time: time.Now()}
+
+	// If there is different status, reason or message update it
+	if condition.Status != status || condition.Reason != reason || condition.Message != message {
+		condition.Status = status
+		condition.Reason = reason
+		condition.Message = message
+		condition.LastTransitionTime = now
+	}
+	condition.LastHeartbeatTime = now
+}
+
+// Find - Check for the existence of a particular condition type in a list of conditions
+func (conditions ConditionList) Find(conditionType ConditionType) *Condition {
+	for i, condition := range conditions {
+		if condition.Type == conditionType {
+			return &conditions[i]
+		}
+	}
+	return nil
+}
+
+// InitCondition - Either return the current condition (if non-nil), or return an empty Condition
+func (conditions ConditionList) InitCondition() *Condition {
+	cond := conditions.GetCurrentCondition()
+
+	if cond == nil {
+		return &Condition{
+			Type:    CommonCondTypeEmpty,
+			Reason:  CommonCondReasonInit,
+			Message: string(CommonCondReasonInit),
+		}
+	}
+
+	return cond.DeepCopy()
+}
+
+// GetCurrentCondition - Get current condition with status == corev1.ConditionTrue
+func (conditions ConditionList) GetCurrentCondition() *Condition {
+	for i, cond := range conditions {
+		if cond.Status == corev1.ConditionTrue {
+			return &conditions[i]
+		}
+	}
+
+	return nil
+}
+
+// UpdateCurrentCondition - update current state condition, and sets previous condition to corev1.ConditionFalse
+func (conditions *ConditionList) UpdateCurrentCondition(conditionType ConditionType, reason ConditionReason, message string) {
+	//
+	// get current condition and update to corev1.ConditionFalse
+	//
+	currentCondition := conditions.GetCurrentCondition()
+	if currentCondition != nil {
+		conditions.Set(
```

Comment:
so I suppose this should keep the historical trail of a given condition type without getting the list longer. So each condition type would only have at most two instances of it in the list: transition src and dst.

Reply:
this is to set the current latest condition to ConditionFalse, if it exists